openwisp_monitoring/device/api/views.py

```diff
@@ -203,6 +205,72 @@ def _write(self, request, pk):
                 metric.write(client['mac'])
             if created:
                 self._create_clients_graph(metric)
+        if 'resources' not in data:
+            return
+        self._write_cpu(data['resources'], pk, ct)
+        self._write_disk(data['resources'], pk, ct)
+        self._write_memory(data['resources'], pk, ct)
+
+    def _write_cpu(self, resources, primary_key, content_type):
+        extra_values = {
+            'load_1': float(resources['load'][0]),
+            'load_5': float(resources['load'][1]),
+            'load_15': float(resources['load'][2]),
+        }
+        metric, created = Metric._get_or_create(
+            object_id=primary_key,
+            content_type=content_type,
+            key='cpu',
+            field_name='cpu_usage',
+            name='Resources: CPU',
+        )
+        metric.write(
+            float(resources['load'][0] / resources['cpus']), extra_values=extra_values,
+        )
+        if created:
+            self._create_resources_graph(metric, resource='cpu')
+            self._create_resources_threshold(metric, resource='cpu')
+
+    def _write_disk(self, resources, primary_key, content_type):
+        used_bytes, size_bytes, available_bytes = 0, 0, 0
+        for disk in resources['disk']:
```

Comment:
here, I think you're not taking into account the case in which the `disk` key in `resources` is not present.
It may happen. If it does happen, it looks to me that the code will fail and the metric collection will fail.
I think you did check this before, was it @PabloCastellano that suggested you to not do it?

Reply:
Yes sometime back https://github.com/openwisp/openwisp-monitoring/pull/76#discussion_r433954476, but I agree with his arguement. @nemesisdesign, the only case in which disk key might not be present is if some other `lua` script is used but then that holds for `memory` and `load` too, right? Shall I change the schema to make these 3 required always or simply revert back to checking if these keys are present and then proceed with the code.